packages/SwingSet/src/kernel/state/vatKeeper.js

```diff
@@ -119,6 +193,12 @@ export function makeVatKeeper(
     return value;
   }
 
+  // We keep the vat's reap-dirt state (and threshold) in a
+  // write-through cache, and populate the cache at vatKeeper startup.
+  const reapDirtKey = `${vatID}.reapDirt`;
+  let reapDirt = JSON.parse(getRequired(reapDirtKey));
+  let reapDirtThreshold = getOptions().reapDirtThreshold;
```

Comment:
I'm having funny feelings about ad-hoc caching, in particular because I don't fully understand the impact of savepoints and crank rollbacks.

Reply:
I probably haven't considered it thoroughly enough, but my basic idea was that crank rollbacks still happen in consensus, and still take time, so some of our reactions to their costs should still happen even if the crank gets rolled back. Like, the runPolicy should be told about computrons spent inside rolled-back cranks (we don't get more time for a block just because the vat did something weird and needs to forget the delivery).

Obviously that doesn't apply to the amount of dirt inside a vat: abandoning the external consequences of the delivery will also abandon the vat-internal changes, like an increase in the amount of dirt. So I agree it would be appropriate to *not* increment `reapDirt` in the case of an abandoned crank.

I wanted to avoid too much extra DB activity for every crank. We definitely need to write out the new dirt value, so it's persistent (and this triggers an `export-data` IAVL change, sigh). We almost certainly shouldn't *read* the dirt threshold each time, since it rarely changes (and only by `setSourceAndOptions`, which happens in its own non-delivery crank, which isn't subject to rollback since there's no vat).

I currently call `addDirt()` from two places (the generic one in `processDeliveryMessage`, and the `gcKrefs` -only one in `processGCMessage`), so if I remove the `reapDirt` cache and do a full DB read+increment+write cycle each time, I'm paying 2 or 4 DB queries per crank. If I can implement your other suggestion to pass `gcKrefs` into `deliveryCrankResults()`, then that is reduced to 2 per crank.

So I think I'll:
* merge gcKrefs into `deliveryCrankResults()`
* remove the RAM cache of `reapDirt`, doing 2 DB actions instead of 1 each delivery
* leave the `reapDirtThreshold` RAM cache

Reply:
Actually I removed all the caches. So on each crank, we do three reads: `vNN.reapDirt`, `kernel.reapDirtThreshold` (the kernel-wide defaults), and `vNN.options` (to get the per-vat overrides). Then we write the updated `vNN.reapDirt`, and occasionally do more if it's time to schedule a BOYD. I agree that this is safer, and the extra DB activity is probably minor compared to what it takes to actually perform the delivery.